gson/src/main/java/com/google/gson/internal/bind/EnumTypeAdapter.java

```diff
@@ -59,17 +59,20 @@ private EnumTypeAdapter(final Class<T> classOfT) {
       // Uses reflection to find enum constants to work around name mismatches for obfuscated
       // classes
       Field[] fields = classOfT.getDeclaredFields();
-      ArrayList<Field> constantFieldsList = new ArrayList<>(fields.length);
+      int constantCount = 0;
       for (Field f : fields) {
+        // Filter out non-constant fields, replacing elements as we go
         if (f.isEnumConstant()) {
-          constantFieldsList.add(f);
+          fields[constantCount++] = f;
         }
       }
 
-      Field[] constantFields = constantFieldsList.toArray(new Field[0]);
-      AccessibleObject.setAccessible(constantFields, true);
+      // Trim the array to the new length
```

Comment:
My initial thought was that most of the time you won't need to do this. Every call to `getDeclaredFields()` returns a new array and if there are no fields that are _not_ enum constants then you can just use that array as is. However, with javac at least, it turns out that there is always a private `$VALUES` field which is an array that is cloned every time `values()` is called on the enum. So perhaps add a further note like "This will always be needed because of the $VALUES field."

Reply:
Good call; comment added.